tools/shared/src/main/scala/org/scalajs/core/tools/io/VirtualFiles.scala

```diff
@@ -122,19 +122,36 @@ trait WritableVirtualJSFile extends WritableVirtualTextFile with VirtualJSFile {
 }
 
 /** A virtual file containing Scala.js IR.
- *
- *  This can be a [[VirtualScalaJSIRFile]] (with [[RelativeVirtualFile]]) or a
- *  [[VirtualFileContainer]].
  *
  *  The main difference compared to using individual files
  *  (that are extracted beforehand) is that the fileset can be versioned at a
  *  higher level: the container needs to change its version when any of the
  *  files change. Therefore, the entire extraction process can be cached.
  */
 trait ScalaJSIRContainer extends VirtualFile {
```

Comment:
Why does this not extend `VirtualFileContainer` and directly define `sjsirFiles`?

Reply:
Because `VirtualScalaJSIRFile extends ScalaJSIRContainer`. So if `ScalaJSIRContainer extends VirtualFile`, then `VirtualScalaJSIRFile` transitively implements `VirtualFileContainer`, and everything goes berserk.

Reply:
I see... LOL. Implementation inheritance really sucks :P